### DISCUSSION TITLE
How to properly add nodes and update positions

### DISCUSSION THREAD

**Discussion by ericjunkins** [xyflow/xyflow#1745] [Q&A]
I'm wanting to have a graph that when only shows initially a depth of 2 (root + first children of root), and if I click on any of the nodes it will add all that nodes children recursively. I made a [codesandbox](https://codesandbox.io/s/fervent-mendeleev-vc4nw?file=/src/ReactFlowGraph.js) to demonstrate the rough thing that I want. There were 2 issues with this.

- Links disappear, which I found a workaround by setting `isHidden = false` on the elements 
- Positions update and stay static based on what has ever been added to the graph, if i close a node section the positions of the children are still being used in the dagre calculation of the positions. 

### Graph with nodes opened
![image](https://user-images.githubusercontent.com/25470708/144911904-4358b61c-823a-4c17-9778-638465513a97.png)

### Graph with nodes closed
![image](https://user-images.githubusercontent.com/25470708/144911965-3b25439d-071d-441d-ab57-7ceae7471769.png)

Ideally this would collapse back into the initial graph without large spaces between the nodes.

@datoslabs hinted in an #1736 that the method that I use to update the `elements` is inadvisable, as it causes many potential issue with race conditions and also inefficient with re renders. Is there a better method to manage my `elements` array?

**Comment by datoslabs** [comment]
FYI, my initial encounter of similar edge rendering issues is documented in #1305.  You could try wrapping `updateNodeDimensions` within a `setTimout` and see if it works for your use case.  I don't like the static timeout value as it may not work for all; if calling the `updateNodeDimensions` after a certain amount of timeout works, I would suggest the following:

```
...
const nodes = useStoreState((s) => s.nodes);
useEffect(()=>{
  if (nodes && nodes.length > 0 && nodes.every(node=>node.__rf.width && node.__rf.height))
    updateNodeDimensions(...);
},[nodes]);
``` 

BTW, I would also suggest wrapping layout logic in similar `useEffect` (see #1639). 

Please also note that V10 contains an overhaul of internal state management API, I am not sure how/if updateNodeDimension would work in V10.

**Reply by ericjunkins** [reply]
I think I'm not fully understanding something here. I tried to look through the `react-flow-renderer` and documentation and examples and can't find any `updateNodeDimensions` function in it, nor in the `dagre` package. Is there something I'm missing for that?

**Reply by datoslabs** [reply]
`updateNodeDimensions` is not part of the documented API therefore subject to change, hence I did not suggest it in my original response.  It is part of the internal state management APIs `react-flow` uses and exported to support more advanced use cases.  `dagre` is an external library used in the [layouting example](https://reactflow.dev/examples/layouting/); it works in the example because the node dimensions are fixed but in real-world usecases where node contents can vary, the comments in the example explains how to get the actual node dimensions to properly layout the graph.  A lot of the information I am sharing are documented in closed issues as well as the source code. 

As a point of clarification, when I said it's not efficient, I meant it's not efficient in the context of large `elements[]` and having to recreate this array in each update cycle.  I would suggest the following:

A. For your usecase, do you want to allow users to DnD the nodes?  If so, performing an auto layout in each update cycle may be confusing for the users as the nodes will move where they don't expect them to; therefore, auto layout should ideally only run during initial rendering (and/or when you offer user an auto layout button) and keep/update elements array.

B. It may be beneficial to make `TestGraph` a tree (instead of array) if you plan to support multiple layers and allow users to collapse/expand at any level, for example:
```
TestGraph = {
  A: { 
    label: "A", 
    type: "input", 
    children:{
      B:{
        label: "B",
        children:{...}
      }
    } 
  }
}
```
I would use the `node path` (e.g. "A.children.B") as the node ID so when showing/hiding nodes, I can quickly reference/traverse the designated branch. 

C. For further optimization, I would construct and maintain an internal state of elements array as well as keep each node reference in a tree:
```
const _elements=[];
const _edges=[];
const _graph={};
//use visitor pattern to traverse tree and do the work where key === node_path
traverse_tree(TestGraph, (node, key)=>{
  const el = {id:key, isHidden=false, ....};
  _elements.push(el);
  add_to_graph({graph:_graph, path=key, node=el});
  ...//build edges
});
```
This way I can directly traverse a specific branch (based on the nodeID/node_path from the `onClick` event) and directly update the branch nodes' `isHidden `property quickly and call `setElements(_elements.concat(_edges))`.

**Comment by ravanliyski** [comment]
I've spent hours going through open and closed issues.
Can someone provide a working example of dynamically changing graph + directed layout?
